src/ssh/test-sshservice.c

```diff
@@ -0,0 +1,801 @@
+
+  const gchar *old_ask;
+} TestCase;
+
+typedef struct {
+  const char *user;
+  const char *password;
+} TestFixture;
+
+static GString *
+read_all_into_string (int fd)
+{
+  GString *input = g_string_new ("");
+  gsize len;
+  gssize ret;
+
+  for (;;)
+    {
+      len = input->len;
+      g_string_set_size (input, len + 256);
+      ret = read (fd, input->str + len, 256);
+      if (ret < 0)
+        {
+          if (errno != EAGAIN)
+            {
+              g_critical ("couldn't read from mock input: %s", g_strerror (errno));
+              g_string_free (input, TRUE);
+              return NULL;
+            }
+        }
+      else if (ret == 0)
+        {
+          return input;
+        }
+      else
+        {
+          input->len = len + ret;
+          input->str[input->len] = '\0';
+        }
+    }
+}
+
+static gboolean
+start_mock_sshd (const gchar *user,
+                 const gchar *password,
+                 GPid *out_pid,
+                 gushort *out_port)
+{
+  GError *error = NULL;
+  GString *port;
+  gchar *endptr;
+  guint64 value;
+  gint out_fd;
+
+  const gchar *argv[] = {
+      BUILDDIR "/mock-sshd",
+      "--user", user,
+      "--password", password,
+      NULL
+  };
+
+
+  g_spawn_async_with_pipes (BUILDDIR, (gchar **)argv, NULL, G_SPAWN_DO_NOT_REAP_CHILD, NULL, NULL,
+                            out_pid, NULL, &out_fd, NULL, &error);
+  g_assert_no_error (error);
+
+  /*
+   * mock-sshd prints its port on stdout, and then closes stdout
+   * This also lets us know when it has initialized.
+   */
+
+  port = read_all_into_string (out_fd);
+  g_assert (port != NULL);
+  close (out_fd);
+  g_assert_no_error (error);
+
+  g_strstrip (port->str);
+  value = g_ascii_strtoull (port->str, &endptr, 10);
+  if (!endptr || *endptr != '\0' || value == 0 || value > G_MAXUSHORT)
+      g_critical ("invalid port printed by mock-sshd: %s", port->str);
+
+  *out_port = (gushort)value;
+  g_string_free (port, TRUE);
+  return TRUE;
+}
+
+static void
+setup_mock_sshd (TestCase *test,
+                 gconstpointer data)
+{
+  const TestFixture *fix = data;
+
+  start_mock_sshd (fix->user ? fix->user : g_get_user_name (),
+                   fix->password ? fix->password : PASSWORD,
+                   &test->mock_sshd,
+                   &test->ssh_port);
+
+  cockpit_ssh_specific_port = test->ssh_port;
+}
+
+static void
+stop_mock_sshd (GPid mock_sshd) {
+  GPid pid;
+  int status;
+
+  pid = waitpid (mock_sshd, &status, WNOHANG);
+  g_assert_cmpint (pid, >=, 0);
+  if (pid == 0)
+    kill (mock_sshd, SIGTERM);
+  else if (status != 0)
+    {
+      if (WIFSIGNALED (status))
+        g_message ("mock-sshd terminated: %d", WTERMSIG (status));
+      else
+        g_message ("mock-sshd failed: %d", WEXITSTATUS (status));
+    }
+  g_spawn_close_pid (mock_sshd);
+}
+
+static void
+teardown_mock_sshd (TestCase *test,
+                    gconstpointer data)
+{
+  stop_mock_sshd (test->mock_sshd);
+
```

Comment:
Do you plan to add something else here still? Seems easier to move  the `stop_mock_sshd()` code into here and only have one teardown function, particulary as it doesn't appear to be completely reentrant (closing the pid if already closed), see below.